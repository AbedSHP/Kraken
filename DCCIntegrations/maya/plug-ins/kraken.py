import sys, json

from PySide import QtGui

import kraken.ui
from kraken.ui.kraken_ui import KrakenUI

from maya import cmds
# import maya.OpenMaya as OpenMaya
import maya.OpenMayaUI as OpenMayaUI
import maya.OpenMayaMPx as OpenMayaMPx

try:
  # Maya 2013 with custom pyside build
  import PySide.shiboken as shiboken
except:
  # Maya 2014 and higher
  import shiboken



# Command
class OpenKrakenEditorCommand(OpenMayaMPx.MPxCommand):
  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  # Invoked when the command is run.
  def doIt(self,argList):
    parent = OpenMayaUI.MQtUtil.mainWindow()
    parent = shiboken.wrapInstance(long(parent), QtGui.QWidget)
    dfgApp = KrakenUI(parent)

  # Creator
  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( OpenKrakenEditorCommand() )


class KrakenUndoableCmd(OpenMayaMPx.MPxCommand):

  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  def isUndoable(self):
    return True

  def doIt(self, argList):
    return 0

  def redoIt(self):
    SingletonHost.getController().redo()
    return 0

  def undoIt(self):
    SingletonHost.getController().undo()
    return 0

  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( KrakenUndoableCmd() )


# Initialize the script plug-in
def initializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.registerCommand( 'openKrakenEditor', OpenKrakenEditorCommand.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:openKrakenEditor' )
    raise

  try:
    mplugin.registerCommand( 'krakenUndoableCmd', KrakenUndoableCmd.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:krakenUndoableCmd' )
    raise

# Uninitialize the script plug-in
def uninitializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.deregisterCommand( 'openKrakenEditor' )
  except:
    sys.stderr.write( 'Failed to unregister command: openKrakenEditor' )


  try:
    mplugin.deregisterCommand( 'krakenUndoableCmd' )
  except:
    sys.stderr.write( 'Failed to unregister command: krakenUndoableCmd' )
