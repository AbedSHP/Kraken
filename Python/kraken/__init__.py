"""Kraken Framework."""

__all__ = ['core', 'examples', 'helpers', 'plugins', 'gui']