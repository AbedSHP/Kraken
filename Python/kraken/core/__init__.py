"""Kraken Core."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "alpha"


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Return:
    String, current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
