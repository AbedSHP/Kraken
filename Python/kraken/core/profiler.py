"""Kraken - objects.profiler module.

Classes:
Profiler - Profiler Object.

"""
import time




class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None

    class ProfilerItem(object):

        def __init__(self, label):
            super(Profiler.ProfilerItem, self).__init__()

            t = time.time()
            self.label = label
            self.start = t
            self.end = t
            self.children = []

        def addChild(self, item):
            self.children.append(item)

        def endProfiling(self):
            self.end = time.time()

    def __init__(self, label='Root'):
        super(Profiler, self).__init__()
        self.reset(label=label)

    def reset(self, label='Root'):
        """Resets the profiler for generating a new report

        Return:
        None

        """
        self.__roots = []
        self.__stack = []

    def push(self, label):
        """Adds a new child to the profiling tree and activates it.

        Return:
        None

        """

        item = Profiler.ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)
        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to the parent item

        Return:
        None

        """
        end = time.time()
        if len(self.__stack) == 0:
            raise Exception("Unable to close bracket. Pop has been called more times than push.") 
             
        self.__stack[-1].endProfiling()
        self.__stack.pop()
        
    def generateReport(self):
        """Returns a json object that encodes the data gathered during profiling.

        Return:
        the json object

        """
        if len(self.__stack) != 0:
            raise Exception("Profiler brackets not closed properly. pop must be called for every call to push. Pop needs to be called another " + str(len(self.__stack)) + " times") 
        report = {
            'tree': None
        }
        functions = {}
        def reportItem(item):
            itemReport  = {
                'label': item.label,
                'duration' : item.end - item.start
            }
            if item.label not in functions:
                functions[item.label] = itemReport['duration']
            else:
                functions[item.label] += itemReport['duration']

            for childItem in item.children:
                if 'children' not in itemReport:
                    itemReport['children'] = []
                itemReport['children'].append(reportItem(childItem))
            return itemReport

        report['tree'] = []
        for rootItem in self.__roots:
            report['tree'].append(reportItem(rootItem))
        return report

    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Return:
        The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
