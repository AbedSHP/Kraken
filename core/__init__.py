"""Kraken Framework."""

VERSION_MAJOR = 0
VERSION_MINOR = 0
VERSION_BUILD = 1
VERSION_SUFFIX = "beta"


def getVersion():
    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString