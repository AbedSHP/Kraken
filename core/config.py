import json

envConfig = {
                "app": "xsi",
                "namingTemplate": {
                                    "suffixes": {
                                                    "null":"null",
                                                    "control":"ctrl",
                                                    "zero":"zero",
                                                    "srt":"srt",
                                                    "geometry":"geo",
                                                    "curve":"crv",
                                                    "visualCue":"visCue",
                                                    "implicit":"imp",
                                                    "hierarchy":"hrc",
                                                    "upVector":"upV",
                                                    "envelope":"env",
                                                    "group":"grp",
                                                    "parameterIn":"paramIn",
                                                    "parameterOut":"paramOut",
                                                    "kineIn":"kineIn",
                                                    "kineOut":"kineOut",
                                                    "srtIn":"srtIn",
                                                    "srtOut":"srtOut",
                                                    "root":"rt",
                                                    "effector":"eff",
                                                    "bone":"bone",
                                                    "iceHost":"iceHost",
                                                    "pointCloud":"pcloud"
                                                },
                                    "locations": {
                                                    "left":"L",
                                                    "medial":"M",
                                                    "right":"R"
                                                }
                                  }
            }


def setEnvConfigFromFile(filePath):
    """Sets the 'envConfig' from a json file.

    Arguments:
    filePath -- String, path on disk of json config file.

    """

    with file(filePath, "r") as configFile:
        config = json.load(filePath)

    global envConfig
    envConfig = config

    return