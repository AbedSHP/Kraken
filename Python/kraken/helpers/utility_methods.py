

def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Return:
    None

    """

    print kObject.getFullName()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)



from kraken.core.maths.math_object import MathObject
from kraken.core.maths import decodeValue

def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData

def prepareToLoad(jsonData):

    return __convertFromJSON(jsonData)

def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertToJSON(value)
    return jsonData

def prepareToSave(jsonData):
    return __convertToJSON(jsonData)