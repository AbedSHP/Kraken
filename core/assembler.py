"""Kraken - Assembler."""

from kraken.core.objects import containers
from kraken.core import *

from collections import OrderedDict
import json


class Assembler(object):
    """Kraken Assembler."""

    def __init__(self):
        super(Assembler, self).__init__()
        self.container = None
        self.definition = {
                        "app":"Kraken",
                        "version":getVersion(),
                        "definition":{}
                      }


    def addContainer(self, container):
        """Adds a container to the Assembler.

        Arguments:
        container -- Container to add to this Assembler.

        Return:
        True if successful.
        """

        if container.name in self.containers.keys():
            raise KeyError("Container is already part of the Assembler: " + container.name)
            return False

        self.containers[container.name] = container
        container.container = self

        return container


    def removeContainer(self, containerName):
        """Removes a container from this Assembler.

        Arguments:
        containerName -- String, name of the container to remove from the Assembler.

        Return:
        True if successful.
        """

        if componentName not in self.components.keys():
            raise KeyError("Container not found in this Assembler: " + componentName)
            return False

        targetContainer = self.components[componentName]
        targetContainer.container = None

        del self.components[componentName]

        return True


    # ================
    # Build Functions
    # ================
    def buildDef(self):
        """Builds the Rig Definition and stores to rigDef attribute.

        Return:
        Dictionary of object data.
        """

        if self.container is None:
            raise AttributeError("Assembler has no container!")

        self.definition["definition"] = self.container.buildDef()

        return self.definition


from kraken.core.objects.containers import *

assembler = Assembler()
container = Container("MyContainer")
assembler.container = container
print json.dumps(assembler.buildDef())