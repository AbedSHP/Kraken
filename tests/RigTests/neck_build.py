
from kraken import plugins
from kraken.examples.neck_component import NeckComponent
from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import logHierarchy
import json


Profiler.getInstance().push("neck_build")

neck = NeckComponent("neck")
logHierarchy(neck)

builder = plugins.getBuilder()
builder.build(neck)

Profiler.getInstance().pop()

if __name__ == "__main__":
    print json.dumps(Profiler.getInstance().generateReport(), sort_keys=False, indent=4, separators=(',', ': '))
else:
    logHierarchy(neck)