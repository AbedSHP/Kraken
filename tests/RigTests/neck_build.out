Warning: Generating 'deformers' layer for component:neck
Failed to find DCC builder. Falling back to Python builder.
neck
neck.inputs
neck.inputs.neckBase
neck.outputs
neck.outputs.neckEnd
neck.outputs.neck
neck.neck
neck.neck.neck
neck.deformers
neck.deformers.neck
